metagraph/tests/test_vector_algorithm.cpp

```diff
@@ -160,6 +160,118 @@ TEST(IntVector, call_nonzeros_sparse_every_4) {
     }
 }
 
+TEST(IntVector, atomic_exchange_and_fetch_sparse) {
+    for (size_t w = 1; w <= 64; ++w) {
+        for (auto memorder : { __ATOMIC_RELAXED, __ATOMIC_SEQ_CST }) {
+            sdsl::int_vector<> vector_atomic = aligned_int_vector(200, 0, w, 16);
+            sdsl::int_vector<> vector(200, 0, w);
+            ASSERT_LT(65u, vector.size());
+            std::atomic<size_t> counter = 0;
+
+            std::mutex mu;
+            std::mutex mu2;
+            #pragma omp parallel for num_threads(3) schedule(dynamic)
+            for (size_t j = 0; j < 3; ++j) {
+                for (size_t i = j; i < vector.size(); i += 3) {
+                    if (i % 130 == 1) {
+                        uint64_t val = (++counter) % w;
+                        EXPECT_EQ(0u, atomic_exchange(vector_atomic, i, val, mu, memorder));
+                        std::lock_guard<std::mutex> lock(mu2);
+                        vector[i] = val;
+                    }
+                }
+            }
```

Comment:
Here each bit is changed only once by a single thread.
Do you also have a test where you write to the same bits?

Reply:
same comment for the other tests

Reply:
Since edits are made to the vector in blocks of variable width (ranging from 1 to 32 bytes), this test will fail if the operations are not atomic.

But you're right, it's also good to have another test where the same positions are written as well.

Reply:
I can also make the modulus smaller to make this test more rigourous.

Reply:
yes, the test should be different this is too simple. Obviously fetch will return 0 if you haven't done any midifications. You need to run something more complex. Otherwise, this test is useless.

Reply:
for instance, you can flip every bit 1001 times by every thread and in the end check that all bits were flipped from 0 to 1.

Reply:
similar for other functions, like fetch_and_add

Reply:
Also, remove all locks from tests where you test atomic operations, they add some synchronization which you don't want.

Reply:
I replaced the tests with those which should affect overlapping indexes in multiple threads.

Reply:
please make sure each of these tests runs less than 1 sec, better 0.1 sec.